scipy/sparse/linalg/eigen/lobpcg/lobpcg.py

```diff
@@ -78,29 +78,41 @@ def _makeOperator(operatorInput, expectedShape):
 
 def _applyConstraints(blockVectorV, factYBY, blockVectorBY, blockVectorY):
     """Changes blockVectorV in place."""
-    gramYBV = np.dot(blockVectorBY.T.conj(), blockVectorV)
-    tmp = cho_solve(factYBY, gramYBV)
+    YBV = np.dot(blockVectorBY.T.conj(), blockVectorV)
+    tmp = cho_solve(factYBY, YBV)
     blockVectorV -= np.dot(blockVectorY, tmp)
 
 
 def _b_orthonormalize(B, blockVectorV, blockVectorBV=None, retInvR=False):
+    normalization = blockVectorV.max(axis=0)
+    blockVectorV = blockVectorV / normalization
     if blockVectorBV is None:
         if B is not None:
             blockVectorBV = B(blockVectorV)
         else:
             blockVectorBV = blockVectorV  # Shared data!!!
-    gramVBV = np.dot(blockVectorV.T.conj(), blockVectorBV)
-    gramVBV = cholesky(gramVBV)
-    gramVBV = inv(gramVBV, overwrite_a=True)
-    # gramVBV is now R^{-1}.
-    blockVectorV = np.dot(blockVectorV, gramVBV)
-    if B is not None:
-        blockVectorBV = np.dot(blockVectorBV, gramVBV)
     else:
+        blockVectorBV = blockVectorBV / normalization
+    VBV = blockVectorV.T.conj() @ blockVectorBV
+    try:
+        # VBV is a Cholesky factor from now on...
+        VBV = cholesky(VBV, overwrite_a=True)
```

Comment:
This is a solve operation. Inverse should be bypassed. You can use chofactor and chosolve routines from scipy.linalg as the commemted part does.

Reply:
> 
> 
> This is a solve operation. Inverse should be bypassed. You can use chofactor and chosolve routines from scipy.linalg as the commemted part does.

I could not figure out how to make it to work, since the inverse must be multiplied on the left, while the chosolve mathematically multiplies by the  inverse of the Cholesky factor on the right only. I have tried chofactor and chosolve with various combinations of .T , but could not get the same answer as currently in the code. For small-size VBV, there is probably no any performance gain anyway by switching to chosolve, but it would still be good to have and try this alternative, I agree. @ilayn If you figure out how to do it, please let me know.

Reply:
Sure I'll have a look in the evening

Reply:
@ilayn If you did not find a better way, should we mark this conversation resolved?

Reply:
In the case where B at least is not identity, it's faster to have an explicit inverse, so that the inversion can be reused. I don't think there's any significant stability difference

Reply:
@antoine-levitt The algorithm needs the inverse of the Cholesky factor of the Gram matrix (computed a few lines below), not of the the Gram matrix  itself.

An alternative could be to compute directly, via SVD, the inverse of the SPD square root of the Gram matrix, which would be very stable, but more expensive for large block sizes. This could be a good next PR, to treat the case when Cholesky fails here.

Reply:
Ah sorry I should have mentioned. Based on your previous comment on the other reply I was waiting for the next PR indeed. I don't want to be a blocker for this. We can do another sweep next if the functionality is sound.

Reply:
@ilayn thanks! Mark "resolved"